contracts/contracts/GrantRoundManager.sol

```diff
@@ -99,66 +126,97 @@ contract GrantRoundManager {
   }
 
   /**
-   * @notice Swap and donate to a grant
-   * @param  _donation Donation being made to a grant
+   * @notice Performs swaps if necessary and donates funds as specified
+   * @param _swaps Array of SwapSummary objects describing the swaps required
+   * @param _deadline Unix timestamp after which a swap will revert, i.e. swap must be executed before this
+   * @param _donations Array of donations to execute
+   * @dev `_deadline` is not part of the `_swaps` array since all swaps can use the same `_deadline` to save some gas
+   * @dev Caller must ensure the input tokens to the _swaps array are unique
+   * @dev Does not verify
    */
-  function swapAndDonate(Donation calldata _donation) external payable {
+  function donate(
+    SwapSummary[] calldata _swaps,
+    uint256 _deadline,
+    Donation[] calldata _donations
+  ) external payable {
     // --- Validation ---
-    // Rounds must be specified
-    GrantRound[] calldata _rounds = _donation.rounds;
-    require(_rounds.length > 0, "GrantRoundManager: Must specify at least one round");
-
-    // Only allow value to be sent if the input token is WETH (this limitation should be fixed in #76, as this
-    // require statement prohibits donating WETH)
-    IERC20 _tokenIn = IERC20(_donation.path.toAddress(0));
-    require(
-      (msg.value == 0 && address(_tokenIn) != WETH) || (msg.value > 0 && address(_tokenIn) == WETH),
-      "GrantRoundManager: Invalid token-value pairing"
-    );
-
-    // Ensure grant recieving donation exists in registry
-    uint96 _grantId = _donation.grantId;
-    require(_grantId < registry.grantCount(), "GrantRoundManager: Grant does not exist in registry");
-
-    // Iterate through each GrantRound to verify:
-    //   - The round has the same donationToken as the GrantRoundManager
-    //   - The round is active
-    for (uint256 i = 0; i < _rounds.length; i++) {
-      require(_rounds[i].isActive(), "GrantRoundManager: GrantRound is not active");
-      require(
-        donationToken == _rounds[i].donationToken(),
-        "GrantRoundManager: GrantRound's donation token does not match GrantRoundManager's donation token"
-      );
+    // TODO consider moving this to the section where we already loop through donations in case that saves a lot of
+    // gas. Leaving it here for now to improve readability
+
+    for (uint256 i = 0; i < _donations.length; i++) {
+      // Validate grant exists
+      require(_donations[i].grantId < registry.grantCount(), "GrantRoundManager: Grant does not exist in registry");
+
+      // Used later to validate ratios are correctly provided
+      donationRatios[_donations[i].token] += _donations[i].ratio;
+
+      // Validate round parameters
+      GrantRound[] calldata _rounds = _donations[i].rounds;
+      require(_rounds.length > 0, "GrantRoundManager: Must specify at least one round");
```

Comment:
I wonder if this is true. While it would possible to donate to a Grant by simply transferring directly to the Payee, we've also discussed that we'd want an event to be emitted even for non-round-associated contributions. So the question is, do we want a separate method for that, one that doesn't swap the donated token? Or do we just use the same method, allowing for no-round donations AND still swapping to the RoundManager's donation token?

Reply:
Good call, I think using the same method and removing this `require` is the way to go since everything else should still behave the same IMO. Going to implement that for now unless anyone objects

Reply:
Cool, I agree, as it will also make the frontend checkout logic must easier as well.